Add CSI Pings for visibility

bradfrizzell opened the pull request:
Work in progress

See https://github.com/ampproject/amphtml/issues/8784

bradfrizzell commented:
@dvoytenko could you take a quick look at what I have done so far and let me know if this is the correct usage (ignoring the fact that I am not actually just sending the pings and just logging for now).

dvoytenko commented on extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js:
```diff
@@ -264,6 +266,53 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
+                        });
+      // 50% vis w ini load
+      vis.listenElement(this.element,
+                        {visiblePercentageMin: 50, waitFor: 'ini-load'},
```
`waitFor: ini-load` is a no-op here. The only thing that matters is the next argument you pass - `whenSignal(ini-load)`. `waitFor` is used by the `amp-analytics` similarly to this upstream.

dvoytenko commented on extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js:
```diff
@@ -264,6 +266,53 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
+      // 50% vis w ini load
+      vis.listenElement(this.element,
+                        {visiblePercentageMin: 50, waitFor: 'ini-load'},
+                        this.signals().whenSignal(CommonSignals.INI_LOAD),
```
This should be `readyPromise`. The trick is that you always need to race ini-promise and load-end, otherwise ini-promise may never show up.

dvoytenko commented on extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js:
```diff
@@ -264,6 +266,53 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
+                        });
+      // ini-load
+      vis.listenElement(this.element, {waitFor: 'ini-load'},
+                        this.signals().whenSignal(CommonSignals.INI_LOAD),
```
ditto: `readyPromise`

dvoytenko commented on extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js:
```diff
@@ -264,6 +266,53 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
+                          console.log(event);
+                        });
+      // load
+      vis.listenElement(this.element, {waitFor: 'load'},
```
No real point to separately catch `load-end`.

dvoytenko commented on extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js:
```diff
@@ -264,6 +266,53 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
+                        });
+      // 50% vis, ini-load and 1 sec
+      vis.listenElement(this.element,
+                        {visiblePercentageMin: 1, waitFor: 'ini-load', totalTimeMin: 1},
```
`totalTimeMin`, et al, are in milliseconds

dvoytenko commented on extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js:
```diff
@@ -206,6 +208,54 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
       insertAnalyticsElement(this.element, this.ampAnalyticsConfig_, true);
     }
+
+    analyticsForDoc(this.getAmpDoc(), true).then(analytics => {
```
Would be probably better to reuse?

dvoytenko commented on extensions/amp-ad/0.1/amp-ad-xorigin-iframe-handler.js:
```diff
@@ -219,6 +220,51 @@ export class AmpAdXOriginIframeHandler {
       }
     });
 
+    if (this.element_.attributes.type.nodeValue == 'adsense' ||
```
`this.element_.getAttribute`. Also, we already have some sort of "google reporter" instance going around. I believe I saw it in many places. I believe we just need to refactor this code into that class and call it from all three places.

bradfrizzell commented:
Addressed your comments in the adsense impl, please take a look and let me know if this is what you meant before I do the same for doubleclick and the xorigin handler.

dvoytenko commented on ads/google/a4a/performance.js:
```diff
@@ -270,4 +272,43 @@ export class GoogleAdLifecycleReporter extends BaseLifecycleReporter {
     return this.initTime_;
   }
+
+  addPingsForVisibility(element, ampDoc, readyPromise) {
```
jsdoc

dvoytenko commented on extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js:
```diff
@@ -264,6 +265,15 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
       insertAnalyticsElement(this.element, this.ampAnalyticsConfig_, true);
     }
+
+    const readyPromise = Promise.race([
```
This formula for `readyPromise` should be fully reusable for any ad (or embed for that matter). I think we should factor it into `addPingsForVisibility`. We'll reduce args and simply this way.

dvoytenko commented on ads/google/a4a/performance.js:
```diff
@@ -270,4 +272,43 @@ export class GoogleAdLifecycleReporter extends BaseLifecycleReporter {
   }
+
+  addPingsForVisibility(element, ampDoc, readyPromise) {
+    analyticsForDoc(ampDoc, true).then(analytics => {
```
Cool. I think this looks great. Let's then use it in the other two cases?

bradfrizzell replied:
sg, will do

dvoytenko commented on ads/google/a4a/performance.js:
```diff
@@ -270,4 +281,56 @@ export class GoogleAdLifecycleReporter extends BaseLifecycleReporter {
+   * @param {!Object} signals Signals object for element.
+   * @override
+   */
+  addPingsForVisibility(element, ampDoc, signals) {
```
Few words about arguments:

1. `ampDoc` is not needed - you can simply pass the `element` to `analyticsForDoc(element, true)`.
2. Why pass `signals`. Isn't this always `element.signals()`?

dvoytenko commented on extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js:
```diff
@@ -206,6 +205,12 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
       insertAnalyticsElement(this.element, this.ampAnalyticsConfig_, true);
     }
+
+    if (this.lifecycleReporter_.addPingsForVisibility) {
```
Why `if (x.addPingsForVisibility)` here but not in the adsense class?

bradfrizzell replied:
removed, don't need it on either

dvoytenko commented on extensions/amp-ad/0.1/amp-ad-xorigin-iframe-handler.js:
```diff
@@ -219,6 +219,14 @@ export class AmpAdXOriginIframeHandler {
       }
     });
 
+    if ((this.element_.attributes.type.nodeValue == 'adsense' ||
```
Why check for `adsense` or `doubleclick` here if `lifecycleReporter` already only implements `addPingsForVisibility` for Google implementation?

bradfrizzell replied:
removed the check

dvoytenko commented:
@bradfrizzell Couple more comments, but otherwise very close to LGTM. LMK.
